BUILDING.md

```diff
@@ -0,0 +1,578 @@
+   cockroach Makefile and the `c-deps/DEP-rebuild` files for details.
+
+Finally, we get to the targets that build the dependencies. The targets are always run and check
+whether the artifacts are out-of-date using the `bin/uptodate` program. If the artifacts are not
+up-to-date, the dependency is built.
+
+The dependency section ends with:
+
+1. setting convenient target names for each dependency (easy to remember and use on the the CLI).
+
+1. a target to run the tests for `libroach`.
+
+
+#### Cockroach Steps
+
+This section starts with cockroach-specific configuration, which includes:
+
+1. Sets TAGS to include `make` and the value of the `native-tag` variable. Export `LC_ALL=C` so all
+   targets produce consistent results (due to sort order).
+
+1. Reset `build/defs.mk.sig` if the build signature changed.
+
+1. Define the resulting binary names for `COCKROACH`, `COCKROACHOSS` and `COCKROACHSHORT`
+
+1. Define target lists for `SQLPARSER_TARGETS`, `PROTOBUF_TARGETS`, `DOCGEN_TARGETS`,
+  `EXECGEN_TARGETS`, `OPTGEN_TARGETS`, `go-targets-ccl`, `go-targets`. It also defines
+  `remove_obsolete_execgen` in there, which removes its namesake files.
+
+Next are (mostly) the targets:
+
+1. The `all` and `c-deps` targets are defined. `c-deps` builds the targets in `C_LIBS_CCL`.
+
+1. Many refinements of environment variable values are made and some dependent targets are defined
+   per-target. This makes it so `go-install`, `${COCKROACH)`, `${COCKROACHOSS)`, and
+   `${COCKROACHSHORT)` all end up running the same recipe but with different parameters to effect
+   their unique outcomes.
+
+    - `build-mode` defines the go command to be run. One of: `build -o TARGET_FILE_NAME`, `install`
+    - `BUILDTARGET` for which target source directory to start from (`./pkg/cmd/cockroach`,
+      `./pkg/cmd/cockroach-oss`, `./pkg/cmd/cockroach-short`)
+    - Adding to `LINKFLAGS` for `$(go-targets)`, `go-install` and the cockroach targets
+
+1. The target and recipe for `go-install`, `${COCKROACH)`, `${COCKROACHOSS)`, and
+   `${COCKROACHSHORT)`.
+
+1. Targets for friendly make target names for `build` -> `${COCKROACH)`,
+   `buildoss` -> `${COCKROACHOSS)`, and `buildshort` -> `${COCKROACHSHORT)`. These targets also:
+
+   - build documentation for the SQL functions and SQL grammar
+   - document the settings used for cross-compipled binaries for `build` and `buildshort`
+
+
+#### Install / Start Cockroach Steps
+
+There are only two things that happen in this section:
+
+- `install`: a target to build and install the `$(COCKROACH)` binary (default location:
+  `/usr/local/bin`)
+- `start`: build and then start an instance of the `$(COCKROACH)` binary
+
+
+#### Cockroach Testing
+
+This section is mostly target and recipe definition for test-related targets. The targets are
+listed here and include notes about them here only if they are not included in `make help` or if
+there is otherwise something interesting to know about the target.
+
+- `testbuild`
```

Comment:
It might be worth expanding `make help` to have some of these -- I had no clue `testbuild` existed, yet it would have been very helpful to me this week.